### PR TITLE
Optimize script summary

### PR DESCRIPTION
**Pull request by Hamms**
Some more improvements to the `script.summarize` method.

For context, this method is extremely demanding in that it individually investigates every stage, script_level, and level within a script to generate this exhaustive summary. To compensate, the method is heavily cached on production; unfortunately, levelbuilder doesn't benefit from this cache and on large scripts this will simply time out.

So this PR includes two optimizations which together reduce the computation time by approximately an order of magnitude on my development machine (from ~1 minute down to ~5.5 seconds).

First, in `stage.cached_script_levels` if we detect there is no cache we simply return the script levels from the database; otherwise, we end up iterating over each level, confirming that it is not in the cache, and then returning it, giving us N useless db queries where one would do just fine.

Second, we add an option to skip the inclusion of prev and next pointers in the level summary; this is a relatively expensive operation, particularly when we're doing it once for _every level_ in a script. We still need this for the individual level summary, but not for the script overview page.

I recommend reviewing [with whitespace changes ignored](https://github.com/code-dot-org/code-dot-org/pull/15785/files?w=1)

### CODE REVIEW

File: dashboard/app/models/script_level.rb [hunk 1]
```diff
@@ -219,7 +219,7 @@ def icon
     end
   end
 
-  def summarize
+  def summarize(include_prev_next=true)
```
Comment (Bjvanminnen) [inline dashboard/app/models/script_level.rb:222]: worth a comment describing this param/when we expect it o be used?

File: dashboard/app/models/stage.rb [hunk 2]
```diff
@@ -192,8 +192,10 @@ def lockable_state(students)
     end
   end
 
-  # Ensures we get the cached ScriptLevels, vs hitting the db.
+  # Ensures we get the cached ScriptLevels if they are being cached, vs hitting the db.
   def cached_script_levels
+    return script_levels unless Script.should_cache?
```
Comment (Bjvanminnen) [inline dashboard/app/models/stage.rb:197]: i wonder if it's misleading that a method called `cached_script_levels` might do an uncached lookup for us. Not sure I can really come up with a better way of expressing this though.

Comment (Hamms) [reply dashboard/app/models/stage.rb:197]: The problem is that without this line, in those situations it's actually doing N uncached lookups :/

Comment (Bjvanminnen) [reply dashboard/app/models/stage.rb:197]: Sure, I understand the need for this line, I'm just wondering if there's a better way we can express things (perhaps by renaming this method? Though I can't think of any good names).

File: dashboard/app/models/script_level.rb (hunk 1 above)
Comment (joshlory) [reply dashboard/app/models/script_level.rb:222]: Is it clear where prev/next are used?  This way we're including them everywhere else to be safe.  I'm curious if they're actually needed.

Comment (Hamms) [reply dashboard/app/models/script_level.rb:222]: It's unfortunately not at all clear to me where they might be used; I am only confident that they are _not_ used in the script overview.

I can in a future PR investigate removing them entirely.

File: dashboard/app/models/stage.rb (hunk 2 above)
Comment (Hamms) [reply dashboard/app/models/stage.rb:197]: `try_cached_script_levels`?

Comment (Bjvanminnen) [reply dashboard/app/models/stage.rb:197]: `try` implies something specific in Ruby land which isn't quite right I think. Sticking with what you have might be fine/best.

**Pull request merged**